daffodil-core/src/test/scala/org/apache/daffodil/dsom/walker/BasicWalker.scala

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+import scala.collection.mutable.ArrayBuffer
+
+class BasicWalker(ignoreTypeWrappers: Boolean = false, onlyElements: Boolean = false) extends AbstractDSOMWalker {
+
+  var nodeArr: ArrayBuffer[Either[TermView, TypeView]] = ArrayBuffer()
+
+  override protected def onWalkBegin(root: RootView): Unit = {}
+
+  override protected def onWalkEnd(root: RootView): Unit = {}
+
+  override def onTermBegin(termElement: TermView): Unit = {
+    if (onlyElements) {
+      termElement match {
+        case _: ElementBaseView => nodeArr += Left(termElement)
+        case _ =>
+      }
+    } else {
+      nodeArr += Left(termElement)
+    }
+  }
+
+  override def onTermEnd(termElement: TermView): Unit = {}
+
+  override def onTypeBegin(typeElement: TypeView): Unit = {
+    if (!ignoreTypeWrappers) {
+      nodeArr += Right(typeElement)
+    }
+  }
+
+  override def onTypeEnd(typeElement: TypeView): Unit = {}
+}
```

Comment:
I'm a little concerned that this ignores a lot of the walker functions. It ignores all End events, and depending on the parameters even more.

Reply:
It ignores all end events because I wanted the end result to be a stack of every element encountered rather than a stack of Nodes with child elements.  In all the implementations so far, End events were just used to remove the current element from the stack, which I don't want here.

Reply:
Understood, but we should still have tests that verify that end events are happening correctly. If there were a bug in the AbstractDSOMWalker that caused certain end events to never happen or to happen out of order, we wouldn't know about it. Whether this test walker does those checks or if it's handled by in each test itself doesn't really seem that important, but something should be able to verify that it's correct.

Reply:
That makes sense, I did overlook that.  I wound up simply adding the element both when the begin and end events are invoked, and then the refactored test for complex types verifies the elements corresponding to those end events occur in the expected indices in the list.